hudi-common/src/main/java/org/apache/hudi/common/util/collection/ExternalSpillableMap.java

```diff
@@ -213,6 +213,8 @@ public R put(T key, R value) {
 
     if (this.inMemoryMap.containsKey(key)) {
       this.inMemoryMap.put(key, value);
+    } else if (inDiskContainsKey(key)) {
+      getDiskBasedMap().put(key, value);
     } else if (this.currentInMemoryMapSize < this.maxInMemorySizeInBytes) {
       this.currentInMemoryMapSize += this.estimatedPayloadSize;
       this.inMemoryMap.put(key, value);
```

Comment:
Any way the record would be put into the map, so why the duplications happened.

Reply:
because currentInMemoryMapSize may be reduced because of size esitimate, as the step bellow:
reproduce:

1、fullfill in-memory map with records, and let this.inMemoryMap.size() % NUMBER_OF_RECORDS_TO_ESTIMATE_PAYLOAD_SIZE == 0

2、insert a record with key1 into ExternalSpillableMap (which will cause size estimate and make sure the currentInMemoryMapSize is still greater than or equal to the maxInMemorySizeInBytes).
it will be spilled to disk.

3、Reduce the size of record of key1 which will make the currentInMemoryMapSize less than maxInMemorySizeInBytes when put into ExternalSpillableMap
it will be put into in-memory map.

Reply:
thank you for your comment. The currentInMemoryMapSize may be **reduced** because of **size esitimate**.  you can see the UT of this PR.  

This is the reproduce steps as bellow:

1、fullfill in-memory map with records （**means that currentInMemoryMapSize  >= maxInMemorySizeInBytes** ）, and **let this.inMemoryMap.size() % NUMBER_OF_RECORDS_TO_ESTIMATE_PAYLOAD_SIZE == 0**

2、insert a record with key1 into ExternalSpillableMap which will trigger size estimate and make sure the currentInMemoryMapSize is still greater than or equal to the maxInMemorySizeInBytes.
it will be spilled to disk.

3、Reduce the size of record of key1 (upsert to ExternalSpillableMap) **which will make the currentInMemoryMapSize less than maxInMemorySizeInBytes** when put into ExternalSpillableMap
it will be put into in-memory map.

this bug also cause data duplication in our production environment.

Reply:
simplify it:
    1、 fullfill  in-memory map first (currentInMemoryMapSize  >= maxInMemorySizeInBytes now )
    2、 insert key1 record into ExternalSpillableMap so it will be spilled to disk
    3、trigger size esitimate make currentInMemoryMapSize  < maxInMemorySizeInBytes again
    4、 upsert key1 record, it will be put into in-memory map because currentInMemoryMapSize  < maxInMemorySizeInBytes

Reply:
@llincc @danny0405 : This shouldn't cause any duplication for get/put queries since the ExternalSpillableMap.get() and ExternalSpillableMap.put() will return/act on latest data. Works because we give precedence to inMemory map compared to diskMap. But the iterator interfaces (values()) will be problematic since ExternalSpillableMap will return duplicate key through this interface. 
 We should maintain the internal constraint that we keep only one copy valid across both in-memory and disk map.

Reply:
@bvaradar @danny0405 yes, the iterator will output all records from memory and disk. But I think it's a easy and clear way to keep only one record  of the same key in the ExternalSpillableMap instead of deduplicate data when iterator.

Reply:
> clear way to keep only one record of the same key in the ExternalSpillableMap instead of deduplicate data when iterator

+1, and I think @bvaradar also have the similiar opinions.

Reply:
Looks good, @bvaradar can you take a look again ~